include/eve/module/real/core/function/regular/simd/x86/fracscale.hpp

```diff
@@ -0,0 +1,142 @@
+//==================================================================================================
+/**
+  EVE - Expressive Vector Engine
+  Copyright : EVE Contributors & Maintainers
+  SPDX-License-Identifier: MIT
+**/
+//==================================================================================================
+#pragma once
+
+#include <eve/detail/implementation.hpp>
+#include <type_traits>
+#include <eve/concept/value.hpp>
+#include <eve/constant/true.hpp>
+
+namespace eve::detail
+{
+  template<floating_real_scalar_value T, typename N>
+  EVE_FORCEINLINE wide<T, N> fracscale_(EVE_SUPPORTS(avx512_), wide<T, N> a0, int s) noexcept
+  {
+    return to_nearest(fracscale)(a0, s);
+  }
+
+  template<int S,floating_real_scalar_value T, typename N>
+  EVE_FORCEINLINE wide<T, N> fracscale_(EVE_SUPPORTS(avx512_), wide<T, N> a0, std::integral_constant<int, S> s) noexcept
+  {
+    return to_nearest(fracscale)(a0, s);
+  }
+
+  // -----------------------------------------------------------------------------------------------
+  // Rouding case
+  template<decorator D, floating_real_scalar_value T, typename N>
+  EVE_FORCEINLINE wide<T, N> fracscale_(EVE_SUPPORTS(avx512_), D const &, wide<T, N> a0, int s) noexcept
+  requires(is_one_of<D>(types<toward_zero_type, downward_type, to_nearest_type, upward_type> {}))
+  {
+    return D()(fracscale[true_(as(a0))])(a0, s);
+  }
+
+  template<int S,decorator D, floating_real_scalar_value T, typename N>
+  EVE_FORCEINLINE wide<T, N> fracscale_(EVE_SUPPORTS(avx512_), D const &, wide<T, N> a0, std::integral_constant<int, S> s) noexcept
+  requires(is_one_of<D>(types<toward_zero_type, downward_type, to_nearest_type, upward_type> {}))
+  {
+    return D()(fracscale[true_(as(a0))])(a0, s);
+  }
+
+  // -----------------------------------------------------------------------------------------------
+  // Masked case
+  template<int S, conditional_expr C, floating_real_scalar_value T, typename N>
+  EVE_FORCEINLINE
+  wide<T, N> fracscale_(EVE_SUPPORTS(avx512_), C const &cx, wide<T, N> const &a0, std::integral_constant<int, S> s) noexcept
+  {
+    return to_nearest(fracscale[cx])(a0, s);
+  }
+
+  template < int R, typename WT> struct  m_imm_choice
```

Comment:
what kind of codegen do we get with this ?

Reply:
however I wrote 2 versions the first wit an integral_constant parameter... If necessary the second one can be suppressed

Reply:
OK, just checked in godbolt:
 - the integral constant is obviously perfect
 - calling with a constant int is perfect too
 - calling with a dynamic integer leads to a lookup table + a jump
 
 I'll try something to see if we can get rid of the jump